### PR TITLE
Add Prereconcile check to PostgreSQL Flexible Server

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Implements the pre-reconciliation check to skip reconciliation of flexible servers that are currently updating.

**Prerequisites**

- [x] PR #2683 

**Special notes for your reviewer**:

Check out the changes in the recordings to see the change in behaviour.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (matthchr) [inline v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:96]: What's the reasoning behind having this check for PostgreSQL FlexibleServer specifically? Is it just showcasing the extension? Did we have a bug where somebody had an issue with the behavior we had previously (is there a corresponding issue?)

Generally speaking the only issue we've had that I'm aware of with blindly PUT-ing and backing off if the resource is in an `-ing` state is with AKS AgentPool, when:
1. Our backoff was to 1m, rather than 3m like now (not a ton of difference I know, but it's something).
2. The AKS AgentPool was stuck in the `-ing` state and would never leave, causing the 1m requests to continue forever. (RP bug).
3. PUT-ing to AKS AgentPool triggers AKS RP to make expensive calls to the Compute RP, which triggered compute throttling. This is also an RP bug.

I like the existence of this extension as a place where we could do GET + diff manually (if we had to, before we have it automated), or if we find a resource that specifically needs the optimization, but I'm not sure what the specific need for it here is?

If there is a specific need - can we leave a comment explaining what it is?

Comment (matthchr) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:96]: On the other hand, if there is no specific need, why bother implementing this extension now? Just leave the hook for if we need it in the future?

Comment (theunrepentantgeek) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:96]: We've had a bug reported for PostgreSQL Flexible server, though I don't recall the exact details and can't find them in my email. But, that's why it's listed in [ADR-2022-12-Reconciliation-Extensions.md](https://github.com/Azure/azure-service-operator/blob/main/docs/hugo/content/design/ADR-2022-12-Reconciliation-Extensions.md) as an affected resource.

Comment (matthchr) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:96]: It feels like we should have some details left in the code as breadcrumbs for "why we bothered to do this for this resource" - for example in this case is the problem due to a bug or bad behavior on the server side? Have we reported it to the PostgreSQL team if so? 

Do you recall how this issue was reported? Through email or otherwise? I can also try to check mine and see if I find anything, if I know more about what I'm looking for.

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (super-harsh) [inline v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:119]: Maybe adding some resource information like `obj.Name` would be helpful here?

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (super-harsh) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:96]: I guess the issue is here #2643

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (matthchr) [inline v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:95]: TBH it might be better to define this as a set of allowed states, rather than a set of disallowed states. ARM defines only [3 terminal states](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/async-api-reference.md#provisioningstate-property)

Those states are:
- Succeeded
- Failed
- Canceled 

Anything other than those 3 states is a transitionary state and will likely cause request reject (that's how AKS works at least, 99% sure most other RPs with "only 1 request active at a time" are the same)

Comment (matthchr) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:95]: This also makes us more hardened against possible futures where they add new "-ing" states in the future (which they are allowed to do)

File: v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go
Comment (matthchr) [inline v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go:261]: Looks like this maybe needs a rebase after the merge of your other PR?

**Review by matthchr — COMMENTED** [review]
Basically LGTM, but waiting to sign-off til after the rebase as PR is a bit hard to fully grok right now as it shows the other PR that has already been merged as well.

File: v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go
Comment (theunrepentantgeek) [reply v2/internal/reconcilers/arm/azure_generic_arm_reconciler_instance.go]: Yup. Done.

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (theunrepentantgeek) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:95]: Fair call. Done.

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (matthchr) [inline v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:95]: minor: fix varname too? This should be `blockingFlexibleServerStates`, right?

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (matthchr) [inline v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:126]: Do we 100% know that `server.Status.State` is not nil? Otherwise we could panic here.

File: v2/api/dbforpostgresql/customizations/flexible_servers_database_extensions.go
Comment (matthchr) [inline v2/api/dbforpostgresql/customizations/flexible_servers_database_extensions.go:39]: Do we 100% know that `server.Status.State` is not nil? Otherwise we could panic here.

**Review by matthchr — APPROVED** [review]

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (theunrepentantgeek) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:95]: No. I changed the name to `nonBlockingFlexibleServerStates` because I changed it to the list of non-blocking states - the states that don't block us from trying to reconcile.

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (theunrepentantgeek) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:126]: Theoretically, `flexibleServerStateBlocksReconciliation()` ensures that `State` is not nil, but to hedge against future issues I refactored anyway.

File: v2/api/dbforpostgresql/customizations/flexible_servers_database_extensions.go
Comment (theunrepentantgeek) [reply v2/api/dbforpostgresql/customizations/flexible_servers_database_extensions.go:39]: Here too.

File: v2/api/dbforpostgresql/customizations/flexible_server_extensions.go
Comment (matthchr) [reply v2/api/dbforpostgresql/customizations/flexible_server_extensions.go:95]: Clearly I was tired at the end of the day

**Pull request merged**
